lib/rspec/core/memoized_helpers.rb

```diff
@@ -123,14 +123,40 @@ def is_expected
 
       # @private
       def __memoized
-        @__memoized ||= {}
+        @__memoized ||= Memoized.new
+      end
+
+      # @private
+      class Memoized
+        def initialize
+          @memoized = {}
+          @mutex    = Mutex.new
+        end
+
+        def for(key, &initializer)
```

Comment:
Capturing a block and using `block.call` is much slower than just using `yield`.  See our benchmark:

https://github.com/rspec/rspec-core/blob/v3.2.0/benchmarks/capture_block_vs_yield.rb

Can you change to `yield`?